config/config.go

```diff
@@ -77,6 +79,9 @@ type Config struct {
 
 	EnableAutoRelay bool
 	StaticRelays    []peer.AddrInfo
+
+	Introspector            coreit.Introspector
+	IntrospectionServerAddr string
```

Comment:
I don't think we should allow the user to pass in an address. For UX, I would recommend the following:

* We add an `ListenAddress() string` method to the `Introspector` interface, so we can query it later.
* The `Introspector` _implementation_ has a `DefaultAddress`, and may use fallback logic if the default listen address is taken. For example, the default address could be: `ws://localhost:8096`, and if the port is occupied, we can fall back to trying 8096 + [1,100], trying one by one until we find an available port, or fail entirely. See https://github.com/libp2p/go-libp2p-daemon/blob/b95e77dbfcd186ccf817f51e95f73f9fd5982600/p2pd/main.go#L30-L60 for an example of this logic.
* We expose the final listen address via the `ListenAddress() string` method.
* Pushing this concern to the implementation is better, because alternate implementations may use different protocols (e.g. Unix sockets) and heuristics.

For a user that wants to set a concrete address, they would construct the concrete implementation via its constructor (which would take a configuration object, that includes the listen addr), and inject that via the `Introspector` option.

Reply:
@raulk 

I am not sure I agree. Why shouldn't we allow to user to pass in a server address while creating a libp2p node ?

The address is a property of the introspection WS server which is owned by the Host. 

The `Introspector` is only  responsible for proving a registry for sub-systems & resolving the state when it is asked to. It is not/shouldn't be responsible for serving that state over the wire & hence dosen't need to be bothered with finding a free port which is anyways not a trivial operation & muddles the construction logic of the `Introspector`.

Reply:
I think I was unclear. We should _totally_ allow passing in a listen address, but not as a dedicated host option. For two reasons:

1. Creating such granular field-level options muddies the namespace.
2. Listen addresses, along with all potential customisations of the introspection endpoint, are implementation-dependent.

Users wanting to customise the listen addr of an introspection server, should be able to do that through the endpoint's constructor:

```go
ep, err := introspection.Endpoint(&introspection.EndpointOpts{
    ListenAddress: "...",
}
h, err := host.New(host.Introspector(ep))
```